### PR TITLE
Implement pytato.concatenate

### PR DESCRIPTION
**Pull request by kaushikcfd**

### CODE REVIEW

**Comment by inducer** [comment]
AFAIK, Github's way of handling "WIP" PRs is by indicating their draft status, either at creation time, or by converting them to draft (on the top right, under "Reviewers")

**Comment by kaushikcfd** [comment]
> AFAIK, Github's way of handling "WIP" PRs is by indicating their draft status, either at creation time, or by converting them to draft (on the top right, under "Reviewers")

Noted and marked as draft.

**Comment by kaushikcfd** [comment]
I don't think I have the permissions to re-trigger the pipeline, but post-https://github.com/inducer/loopy/pull/157 tests should pass.

**Comment by inducer** [comment]
Added you, reran the jobs.

**Review by inducer — COMMENTED** [review]
Looks good, just a few small things.

File: test/test_codegen.py [hunk 1]
```diff
@@ -405,6 +405,30 @@ def test_stack(ctx_factory, input_dims):
...
+    x1 = pt.make_data_wrapper(namespace, x1_in)
+    x2 = pt.make_data_wrapper(namespace, x2_in)
+
+    prog = pt.generate_loopy(
+            pt.concatenate((x0, x1, x2), axis=1),
+            target=pt.PyOpenCLTarget(queue))
+
+    _, (out,) = prog()
+    expected_out = np.concatenate((x0_in, x1_in, x2_in), axis=1)
+    assert (out == expected_out).all()
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply test/test_codegen.py]: How about 0b6d66a?

File: pytato/array.py [hunk 2]
```diff
@@ -965,6 +967,51 @@ def shape(self) -> ShapeType:
 # }}}
 
 
+# {{{ concatenate
+
+class Concatenate(Array):
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:972]: ccb0177785609e67c71c3f8f879f3bf9ab806637

File: pytato/array.py [hunk 3]
```diff
@@ -1386,6 +1433,38 @@ def stack(arrays: Sequence[Array], axis: int = 0) -> Array:
...
+       >>> pt.concatenate(arrays, axis=0).shape
+       (12,)
+
+    :param arrays: a finite sequence, each of whose elements is an
+        :class:`Array` . The arrays are of the same shape except along the
+        *axis* dimension.
+    :param axis: The axis along which the arrays will be concatenated.
+    """
+
+    if not arrays:
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:1451]: 12b87bf1fca539f699f8fe7318e7fac2ad66df4a

File: pytato/array.py [hunk 4]
```diff
@@ -965,6 +967,51 @@ def shape(self) -> ShapeType:
...
 
+# {{{ concatenate
+
+class Concatenate(Array):
+    """Join a sequence of arrays along an existing axis.
+
+    .. attribute:: arrays
+
+        The sequence of arrays to join. The arrays must have same shape except
+        for the dimension corresponding to *axis*.
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py]: 12b87bf1fca539f699f8fe7318e7fac2ad66df4a

File: test/testlib.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import pyopencl as cl
+import numpy
+import pytato as pt
+from pytato.transform import Mapper
+from pytato.array import (Array, Placeholder, MatrixProduct, Stack, Roll,
+                          AxisPermutation, Slice, DataWrapper, Reshape,
+                          Concatenate, Namespace)
+
+
+class EagerEvaluator(Mapper):
```
Comment (inducer) [inline test/testlib.py:11]:
```suggestion
class NumpyBasedEvaluator(Mapper):
```

File: test/testlib.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+def assert_allclose_to_numpy(expr: Array, queue: cl.CommandQueue,
+                              parameters: Dict[Placeholder, Any] = {}):
+    """
+    Raises an :class:`AssertionError`, if there is a discrepancy between *expr*
+    evaluated lazily via :mod:`pytato` and eagerly via :mod:`numpy`.
+
+    :arg queue: An instance of :class:`pyopencl.CommandQueue` to which the
+        generated kernel must be enqueued.
+    """
+    np_result = EagerEvaluator(numpy, expr.namespace, parameters)(expr)
```
Comment (inducer) [inline test/testlib.py:63]:
```suggestion
    np_result = NumpyBasedEvaluator(numpy, expr.namespace, parameters)(expr)
```

**Review by inducer — APPROVED** [review]
One rename, and we're good to go. Thanks for working on this!

**Pull request merged**